### PR TITLE
Proof of Blakers-Massey Theorem

### PR DESCRIPTION
**Pull request by kangrongji**
This PR contains a cubical proof of the Blakers-Massey Theorem and adds some stuff to library that used in the proof. 
It seems there is still no such proof of this fundamental result in homotopy theory, so I decided to spend some time to write one and tried to use as much explicit cubes as possible.

### CODE REVIEW

File: Cubical/Homotopy/BlakersMassey.agda [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+        (transportRefl _ ∙ (λ k → (q₀₀ , λ i j → transpLeftCode-filler refl q i j k))) r
+
+    transpLeftCodeβ :
+        (y : Y) → (q : Q x₀ y) → (q' : fiberSquare q₀₀ q₀₀ refl refl)
+      → transpLeftCode y q ∣ q₀₀ , q' ∣ₕ ≡ ∣ q₀₀ , (λ i j → transpLeftCode-filler (push q) q' i j i1) ∣ₕ
+    transpLeftCodeβ y q q' = transportTrunc _ ∙ (λ i → ∣ transpLeftCodeβ' _ q' i ∣ₕ)
+
+    transpPushCodeβ :
+        (y : Y) → (q : Q x₀ y) → (q' : fiberSquare q₀₀ q₀₀ refl refl)
+      → transport (λ i → pushCode q i (λ j → push q (i ∧ j))) ∣ q₀₀ , q' ∣ₕ ≡ ∣fiber→[q₀₀=q₁₀]∣ q₀₀ q (push q) (λ i j → transpLeftCode-filler (push q) q' i j i1)
```
Comment (mortberg) [inline Cubical/Homotopy/BlakersMassey.agda:616]: Very long line

File: Cubical/Homotopy/BlakersMassey.agda [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+    → {y₀ : Y}(q₀₀ : Q x₀ y₀) → {x₁ : X}(q₁₁ : Q x₁ y₁)
+    → (r : inl x₀ ≡ inr y₁)
+    → (p : push q₀₁ ≡ r)
+    → left→rightCodeExtended q₀₀ q₁₁ r (fiber← q₀₁ q₀₀ q₁₁ r p) ≡ ∣ q₀₁ , p ∣ₕ
+  fiber←→ {x₀ = x₀} {y₁ = y₁} q₀₁ q₀₀' q₁₁' =
+    WedgeConnectivity.extension m n
+      (leftFiber  x₀ , (y₁ , q₀₁)) (leftConn  x₀)
+      (rightFiber y₁ , (x₀ , q₀₁)) (rightConn y₁)
+      (λ (y₀ , q₀₀) (x₁ , q₁₁) →
+        (((r : inl x₀ ≡ inr y₁) → (p : push q₀₁ ≡ r) → left→rightCodeExtended q₀₀ q₁₁ r (fiber← q₀₁ q₀₀ q₁₁ r p) ≡ ∣ q₀₁ , p ∣ₕ )
```
Comment (mortberg) [inline Cubical/Homotopy/BlakersMassey.agda:517]: Long line

File: Cubical/Homotopy/BlakersMassey.agda [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+    → {x₀ : X}(q₀₀ : Q x₀ y₀) → {y₁ : Y}(q₁₁ : Q x₁ y₁)
+    → (r : inl x₀ ≡ inr y₁)
+    → (p : fiberSquarePush q₀₀ q₁₀ q₁₁ r)
+    → right→leftCodeExtended q₀₀ q₁₁ r (fiber→ q₁₀ q₀₀ q₁₁ r p) ≡ ∣ q₁₀ , p ∣ₕ
+  fiber→← {x₁ = x₁} {y₀ = y₀} q₁₀ q₀₀' q₁₁' =
+    WedgeConnectivity.extension m n
+      (leftFiber  x₁ , (y₀ , q₁₀)) (leftConn  x₁)
+      (rightFiber y₀ , (x₁ , q₁₀)) (rightConn y₀)
+      (λ (y₁ , q₁₁) (x₀ , q₀₀) →
+        (((r : inl x₀ ≡ inr y₁) → (p : fiberSquarePush q₀₀ q₁₀ q₁₁ r) → right→leftCodeExtended q₀₀ q₁₁ r (fiber→ q₁₀ q₀₀ q₁₁ r p) ≡ ∣ q₁₀ , p ∣ₕ )
```
Comment (mortberg) [inline Cubical/Homotopy/BlakersMassey.agda:454]: Long line

File: Cubical/Homotopy/BlakersMassey.agda [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+      fiber←→[q₀₀=q₀₁] : fiber→[q₁₁=q₁₀] q₁₁ q₀₁ r (fiber←[q₀₀=q₀₁] q₀₁ q₁₁ r p .snd) .snd ≡ p
+      fiber←→[q₀₀=q₀₁] i j k = fiber←→[q₀₀=q₀₁]-filler i j k i1
+
+    {- q₀₀ = q₀₁ = q₁₁ -}
+    fiber←→hypercube :
+        (r : inl x₀ ≡ inr y₁)
+      → (p : push q₀₁ ≡ r)
+      → PathP (λ i → fiber→[q₀₀=q₁₀=q₁₁] q₀₁ i r (fiber←[q₀₀=q₀₁=q₁₁] q₀₁ i r p .snd) .snd ≡ p)
+              (fiber←→[q₁₁=q₀₁] q₀₁ r p) (fiber←→[q₀₀=q₀₁] q₀₁ r p)
+    fiber←→hypercube r p i j u v =
```
Comment (mortberg) [inline Cubical/Homotopy/BlakersMassey.agda:336]: Wow, this is pretty wild!

File: Cubical/Homotopy/BlakersMassey.agda [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+    → inl x₀ ≡ inr y₁ → Type ℓ
+  fiberSquarePush q₀₀ q₁₀ q₁₁ = fiberSquare q₀₀ q₁₀ (push q₁₁)
+
+  fiber' : {x₀ : X}{y₀ : Y}(q₀₀ : Q x₀ y₀){x₁ : X}{p : Pushout} → inl x₁ ≡ p → inl x₀ ≡ p → Type ℓ
+  fiber' {y₀ = y₀} q₀₀ {x₁ = x₁} r' r = Σ[ q₁₀ ∈ Q x₁ y₀ ] fiberSquare q₀₀ q₁₀ r' r
+
+  fiber'Push : {x₀ x₁ : X}{y₀ y₁ : Y}(q₀₀ : Q x₀ y₀)(q₁₁ : Q x₁ y₁) → inl x₀ ≡ inr y₁ → Type ℓ
+  fiber'Push q₀₀ q₁₁ = fiber' q₀₀ (push q₁₁)
+
+  leftCodeExtended : {x₀ : X}{y₀ : Y}(q₀₀ : Q x₀ y₀)(x₁ : X){p : Pushout} → inl x₁ ≡ p → inl x₀ ≡ p → Type ℓ
```
Comment (mortberg) [inline Cubical/Homotopy/BlakersMassey.agda:79]: Long line

File: Cubical/Homotopy/BlakersMassey.agda [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+Also the HoTT-Agda formalization by Favonia:
+  (https://github.com/HoTT/HoTT-Agda/blob/master/theorems/homotopy/BlakersMassey.agda)
+
+Using cubes explicitly as much as possible.
+
+-}
+{-# OPTIONS  --safe #-}
+module Cubical.Homotopy.BlakersMassey where
+
+open import Cubical.Core.Everything
```
Comment (mortberg) [inline Cubical/Homotopy/BlakersMassey.agda:19]: Not needed as Prelude exports this

**Review by mortberg — CHANGES_REQUESTED** [review]
Very nice stuff! Thanks for contributing 

I only had a few very minor comments about long lines and then I'll be very happy to merge

File: Cubical/Foundations/Prelude.agda [hunk 7]
```diff
@@ -353,6 +354,18 @@ Square :
   → Type _
 Square a₀₋ a₁₋ a₋₀ a₋₁ = PathP (λ i → a₋₀ i ≡ a₋₁ i) a₀₋ a₁₋
 
+-- vertical composition of squares
```
Comment (mortberg) [inline Cubical/Foundations/Prelude.agda:357]: @aljungstrom Didn't you need something like this in your very big PR?

**Review by ecavallo — CHANGES_REQUESTED** [review]

File: Cubical/Homotopy/BlakersMassey.agda (hunk 4 above)
Comment (kangrongji) [reply Cubical/Homotopy/BlakersMassey.agda:336]: haha not my fault. I was forced by the laws of nature.

**Review by mortberg — APPROVED** [review]
Thanks for fixing the long lines so fast! I'll merge as soon as the CI is finished

**Comment by mortberg** [comment]
Oops: 
```
Not in scope:
  isOfHLevelΠ₂
  at /home/runner/work/cubical/cubical/Cubical/Homotopy/BlakersMassey.agda:354,11-23
  ...
```

**Comment by kangrongji** [comment]
> Oops:
> 
> ```
> Not in scope:
>   isOfHLevelΠ₂
>   at /home/runner/work/cubical/cubical/Cubical/Homotopy/BlakersMassey.agda:354,11-23
>   ...
> ```

My bad. Forgot to use new name.

**Review by ecavallo — APPROVED** [review]

**Pull request merged**